Fix Archive failures when using header_mappings_dir and Catalyst

tgoyne opened the pull request:
Catalyst uses the macOS-style versioned framework layout, so it needs the same workaround for header_mappings_dir as macOS frameworks do. We don't know at install time if a pod will be used with Catalyst, so I made it always add the build step for iOS targets and adjusted the script to do nothing if it's running on a non-versioned framework.

I updated the things added in https://github.com/CocoaPods/CocoaPods/pull/5352 to also test the equivalent thing for iOS, but there doesn't appear to be any existing tests which verify that the end result can actually be archived. I manually tested it on both the example project and on Realm, but is there anything further I should do here test-wise?

joeldrotleff commented:
I'm just commenting to bump this up - as mentioned above this issue is causing gRPC builds to fail when *archiving* for Catalyst, which was a major surprise for me trying to port our Firebase app over from iOS to macOS (It was working great in development, then lo and behold the archive step failed).

So very grateful for figuring this out and looking forward to when it's merged!

appfrilans commented:
I'm also super interested in this one! 🙌

paulb777 commented:
@tgoyne It's great to see progress on this long-standing Catalyst issue. Since this is a bugfix, you may want to consider rebasing to the 1-10-stable branch. Also the CI needs to be made green.

tgoyne commented:
Is there a guide to updating the integration specs? The CI failures are just that the generated project has changed.

paulb777 commented:
`bundle exec rake spec:rebuild_integration_fixtures` will rebuild the integration tests, then you can make a PR to the integration test git submodule, and update the submodule reference to the integration test git hash in this repo.

tgoyne commented:
I've retargeted this to stable-1-10 and updated the integration tests (and manually verified that the changes there look sensible to me). I had to bump up the maximum block length in the linter settings since the suite I added the new tests to was already the longest one, so I hope that's okay.

dnkoutso commented on CHANGELOG.md:
```diff
@@ -14,6 +14,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 * None.  
```
nit: remove this since you are adding a first fix!

paulb777 reviewed (approved):
LGTM

appfrilans commented:
So excited about this one! Great work! Any ETA on public release?

dnkoutso commented on spec/unit/installer/xcode/pods_project_generator/pod_target_installer_spec.rb:
```diff
@@ -1391,6 +1404,18 @@ class PodsProjectGenerator
                   build_phase.should.not.be.nil
                 end
 
+                it 'creates a build phase to symlink header folders on iOS' do
```
should this be `macOS`?

tgoyne replied:
The description was correct and the test was wrong. There's already macOS tests for this that I copied and pasted for iOS, and in this spot I failed to actually change the test to iOS.

tgoyne commented:
The new CI failures appear to be due to the CI environment changing rather than anything related to the PR.

dnkoutso commented on CHANGELOG.md:
```diff
@@ -12,7 +12,9 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
-* None.  
+* Fix errors when archiving a Catalyst app which depends on a pod which uses header_mappings_dir.  
+  [Thomas Goyne](https://github.com/tgoyne)
+  [#10016](https://github.com/CocoaPods/CocoaPods/pull/10016)
```
this link is pointing to the wrong PR. Can you update it to 10224?

dnkoutso commented:
@paulb777 havent checked on CI, is it a flake?

paulb777 commented:
Doesn't seem related to this PR. Restarted CI

paulb777 commented:
@dnkoutso  Yep, they were flakes

dnkoutso commented:
@paulb777 I think CHANGELOG needs updating still and I can land it.

appfrilans commented:
Any news on when 1.10.2 is about to be released?

appfrilans commented:
Is Cocoapods not developed anymore? Have not seen any releases and no new progress on the issues for 1.10.2 in months.

igor-makarov commented:
@appfrilans CocoaPods is maintained by volunteers in their free time. Because of this, there's no fixed release schedule.

appfrilans commented:
> @appfrilans CocoaPods is maintained by volunteers in their free time. Because of this, there's no fixed release schedule.

I get that. But perhaps releases could be done more often? Finished (in this case major) fixes are now delayed several months to come out.

shmakovigor commented:
Any ideas on ETA for 1.10.2 with Catalyst issue fix?

varun-pm-tact commented:
Any ETA for 1.10.2?

appfrilans commented:
Wonders the same!

lukemmtt commented:
Would really love to see 1.10.2 released with this fix, or at least as a beta I could install. Thanks very much to @tgoyne for putting together a fix!

willm132 commented:
Any beta I can install to fix this issue?

lukemmtt commented:
> Any beta I can install to fix this issue?

I haven’t tried this myself (since I found a workaround for my issue), but it looks like you could install the master branch of CocoaPods directly using this guide if needed:
https://guides.cocoapods.org/using/unreleased-features

dnkoutso commented:
Going to ship 1.10.2 in a few days.
